src/sqlfluff/parser/grammar.py

```diff
@@ -443,7 +443,7 @@ def _look_ahead_match(cls, segments, matchers, parse_context, code_only=True):
                         seg_buff = seg_buff[1:]
 
     @classmethod
-    def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context, code_only=True):
+    def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context, code_only=True, start_bracket=None, end_bracket=None):
```

Comment:
I agree this function has become pretty long and complex. If I were to step through it in a debugger, I could probably figure out what it does. Some possible "low-hanging fruit" for simplifying it:
* Extract some helper functions to make the function shorter. 
* There are a lot of local variables in this function.
  * This much state suggests maybe organizing the code into a class, e.g. `BracketMatcher`?  If not that, at minimum I'd add some comments to the most important variables, explaining what they contain and what that's used for. Maybe consider adding type annotations in this area of the code.